### ISSUE TITLE
Multiple algorithms for single sets of domains (by default)

### ISSUE THREAD

**Issue by lukas2511** [lukas2511/dehydrated#499]
It would be great if dehydrated would (ideally by default) allow for a multi-cert configuration with RSA and elliptic curves like suggested in the [Let's Encryption Integration Guide](https://letsencrypt.org/docs/integration-guide/). This would allow for modern and fast cryptography with a fallback for older clients.

This would basically affect the `sign_domain` and `command_sign_domain` functions, a few filenames would change but backwards compatibility will be a priority.

I think the ideal output would be a directory structure like this:

```
certs/example.com/cert.rsa.pem -> cert-1234.rsa.pem
certs/example.com/fullchain.rsa.pem -> fullchain-1234.rsa.pem
...
certs/example.com/cert.secp384r1.pem -> cert-1234.secp384r1.pem
certs/example.com/fullchain.secp384r1.pem -> fullchain-1234.secp384r1.pem
...
certs/example.com/cert.pem -> cert-1234.rsa.pem
certs/example.com/fullchain.pem -> fullchain-1234.rsa.pem
```

I'm thinking about having the config look something like this: `KEY_ALGO="rsa secp384r1"`

The main symlinks like `fullchain.pem` would point to the first given algorithm, which will make the new scheme fully compatible with the old one (where only one was set, default being "rsa").

The certificates will be generated fully independent of each other, but double-authorization of domains shouldn't be required as the authorizations are "cached" by the CA, so this wouldn't really make the script much slower and for bigger setups there always will be the option to just choose an algorithm by setting the config variable.

**Comment by darix** [comment]
+1 :D

I already checked the docs. nginx and apache can do it. I will find out if haproxy can do it soon.

**Comment by twrist** [comment]
Haproxy can do it perfectly, provided it's named like `example.com.pem.rsa example.com.pem.ecdsa`

**Comment by lukas2511** [comment]
@twrist great to hear that people are actually using setups like this. does haproxy really need those file-endings or are they configurable?

**Comment by twrist** [comment]
For doing bundling, the file endings are required and cannot be configured to be different. However, if a setup only uses an RSA/ECDSA certificate it doesn't require the extension.

**Comment by alainwolf** [comment]
Nginx can do it, but you can't use it with "`ssl_stapling_file`". As Nginx allows multiple "`ssl_certificate`" statements but only allows one stapling file.

**Comment by martin21** [comment]
Postfix can do it as well. I intend to use it with Postfix and look forward to this feature.

**Comment by mark9064** [comment]
would love to see this as well +1

**Comment by fei0316** [comment]
+1 would love to see that happening. Though ECDSA is quite ubiquitous nowadays, it's always nice to support RSA as well.
As a side question, what would be an appropriate way to implement multiple algorithms for the one set of domains now?

**Comment by heraklit256** [comment]
This would be especially useful for public mail servers, as modern clients could benefit from ECDSA performance, while legacy ones will still be able to use RSA and won't fall back to plaintext.

**Comment by nicoduck** [comment]
Any update on that? I'm trying to figure out the least-painful way to get ECDSA and RSA certs into postfix.

**Comment by darix** [comment]
well the manual way already works

https://nordisch.org/posts/advanced-dehydration/

this issue is about doing it automatically.

**Comment by saz** [comment]
I'm using a simple wrapper script:

/usr/local/bin/dehydrated:
```
#!/bin/bash
for arg do
  shift
  [ "$arg" = "-a" ] && echo 'Do not set `-a`!' && exit 1
  [ "$arg" = "-o" ] && echo 'Do not set `-o`!' && exit 1
  set -- "$@" "$arg"
done

/usr/local/bin/dehydrated.upstream -a rsa -o /etc/dehydrated/certs_rsa "$@"
/usr/local/bin/dehydrated.upstream -a prime256v1 -o /etc/dehydrated/certs_ecc "$@"
```

The original dehydrated command is located in `/usr/local/bin/dehydrated.upstream`

It's still not perfect. E.g. challenge validation might fail and dehydrated might continue (depending on the options you're setting) and the second time it will try it again (as it's executed twice) and you might hit some rate limits.
But it's still easier for me to handle certificates like that.

HTH

**Comment by darix** [comment]
tbh the config file version which is supported by dehydrated out of the box is way less painful than that :)

**Comment by saz** [comment]
@darix can you provide more details?

**Comment by rjhenry** [comment]
Did you read the linked blog post ([here](https://nordisch.org/posts/advanced-dehydration/)? That goes into detail on how to set up multiple-algorithm certificates.

**Comment by saz** [comment]
Yes, but I still don't understand why it's less painful. I need to add the wrapper once, add a domain to domains.txt and that's it.

The other requires a cert specific config for all domains. Looks more painful to me, but that's just my opinion on that :)
